src/core/tests/test_views.py

```diff
@@ -97,3 +125,360 @@ def test_edit_profile(self):
+            response.content.decode(),
+        )
+
+    def test_forgot_password_link_has_next(self):
+        get_data = {
+            'next': self.next_url_raw,
+        }
+        orcid_registration_path = f'/register/step/orcid/{self.orcid_token_uuid}/'
+        response = self.client.get(orcid_registration_path, get_data)
+        self.assertIn(
+            f'/reset/step/1/?next={self.next_url_encoded}',
+            response.content.decode(),
+        )
+
+    def test_register_link_has_next(self):
+        get_data = {
+            'next': self.next_url_raw,
+        }
+        orcid_registration_path = f'/register/step/orcid/{self.orcid_token_uuid}/'
+        response = self.client.get(orcid_registration_path, get_data)
+        self.assertIn(
+            f'/register/step/1/{self.orcid_token_uuid}/?next={self.next_url_encoded}',
+            response.content.decode(),
+        )
+
+
+class ActivateAccountTests(CoreViewTestsWithData):
+
+    @patch('core.views.models.Account.objects.get')
+    def test_activate_success_redirects_with_next(self, objects_get):
+        objects_get.return_value = self.user
+        post_data = {}
+        url = f'/register/step/2/12345/?next={self.next_url_encoded}'
+        response = self.client.post(url, post_data, follow=True)
+        self.assertIn((self.core_login_with_next, 302), response.redirect_chain)
+
+    @patch('core.views.models.Account.objects.get')
+    def test_login_link_has_next(self, objects_get):
+        objects_get.return_value = None
+        get_data = {
+            'next': self.next_url_raw,
+        }
+        response = self.client.get('/register/step/2/12345/', get_data)
+        self.assertIn(
+            self.core_login_with_next,
+            response.content.decode(),
+        )
+
+
+class ReturnURLTests(CoreViewTestsWithData):
+    """
+    These tests check that the url_with_return
+    template tag is present in public-facing templates where
+    the user has an option to log in or register.
+    """
+
+    def test_journal_homepage_account_links_have_return(self):
+        for theme in self.themes:
+            response = self.client.get('/', data={'theme': theme})
+            content = response.content.decode()
+            self.assertIn('/login/?next=/', content)
+            self.assertNotIn('"/login/"', content)
+            self.assertIn('/register/step/1/?next=/', content)
+            self.assertNotIn('"/register/step/1/"', content)
+
+    def test_journal_submissions_account_links_have_return(self):
+        for theme in self.themes:
+            response = self.client.get('/submissions/', data={'theme': theme})
+            content = response.content.decode()
+            self.assertIn('/login/?next=/submissions/', content)
+            self.assertNotIn('"/login/"', content)
+            self.assertIn('/register/step/1/?next=/submissions/', content)
+            self.assertNotIn('"/register/step/1/"', content)
```

Comment:
What are we testing here? that the current template injects its own path onto the login link of the submissions page?

We are not setting a `SERVER_NAME` on the `self.client.get` call, so the result of this test is at the mercy of the state defined in the `settings.py` file. In order to test the journal page explicitly, we need to either use the journal domain as `SERVER_NAME` or run the request in path mode.

Reply:
Yes, that's what we're testing. Without the next parameter, the user would lose their submission context.

Good point nabout the settings--I've updated that.

I've also removed the negative assertions, because I've also reverted the login and registration links in the nav to preserve the old dashboard redirect.